packages/browser-wallet/src/popup/store/settings.ts

```diff
@@ -27,15 +27,49 @@ const storedNetworkConfigurationAtom = atomWithChromeStorage<NetworkConfiguratio
     ChromeStorageKey.NetworkConfiguration,
     mainnet
 );
-export const networkConfigurationAtom = atom<NetworkConfiguration, NetworkConfiguration>(
+export const networkConfigurationAtom = atom<NetworkConfiguration, NetworkConfiguration, Promise<void>>(
     (get) => get(storedNetworkConfigurationAtom),
-    (_, set, networkConfiguration) => {
-        set(storedNetworkConfigurationAtom, networkConfiguration);
-        popupMessageHandler.broadcast(EventType.ChainChanged, networkConfiguration.genesisHash);
-        set(selectedIdentityIndexAtom, 0);
-        storedCredentials
-            .get(networkConfiguration.genesisHash)
-            .then((creds) => set(selectedAccountAtom, creds?.length ? creds[0]?.address : undefined));
+    async (_, set, networkConfiguration) => {
+        const credentials = await storedCredentials.get(networkConfiguration.genesisHash);
+
+        const networkPromise = set(storedNetworkConfigurationAtom, networkConfiguration);
+        const identityPromise = set(selectedIdentityIndexAtom, 0);
+
+        const selectedAccount = credentials?.length ? credentials[0]?.address : undefined;
+        const accountPromise = set(selectedAccountAtom, selectedAccount);
+
+        // Wait for all the derived state of a network change to be done before broadcasting
+        await Promise.all([networkPromise, identityPromise, accountPromise]);
+
+        const connectedSites = await storedConnectedSites.get();
+        const sortedConnectedSites = connectedSites
+            ? Object.entries(connectedSites).sort(([accountA], [accountB]) => {
+                  if (credentials === undefined) {
+                      return 0;
+                  }
+                  return (
+                      credentials.findIndex((c) => c.address === accountA) -
+                      credentials.findIndex((c) => c.address === accountB)
+                  );
+              })
+            : undefined;
+
+        popupMessageHandler.broadcast(EventType.ChainChanged, networkConfiguration.genesisHash, {
+            requireWhitelist: false,
+            nonWhitelistedTabCallback: (tab) => {
+                if (!tab.url) {
+                    return;
+                }
+
+                // If tab has any account connected, send account changed event, otherwise account disconnected.
+                const [firstConnectedAccount] = sortedConnectedSites?.[0] ?? [];
```

Comment:
I'm not sure I understand how this checks if the tab is connected, does it not just check whether there are any site connected?

Reply:
Fixed this now.

Reply:
It looks like it should work, but if I switch between mainnet and testnet, I never get "account disconnected" :thinking:

Reply:
Yeah sorry.. Must have not tested properly here. Should work now.